nameresolution/mdns/mdns.go

```diff
@@ -258,58 +321,168 @@ func (m *resolver) ResolveID(req nameresolution.ResolveRequest) (string, error)
-	// Due to the asynchronous nature of cancel() there
-	// is no guarantee that this will ONLY be invoked on the
-	// first address. Ensure that multiple invocations of this
-	// function are safe.
-	onFirst := func(ip string) {
-		addr = ip
-		cancel() // cancel to stop browsing.
-	}
+		err := m.browse(ctx, appID, onFirst)
+		if err != nil {
+			m.pubErrToSubs(appID, err)
+			return
+		}
 
-	m.logger.Debugf("Browsing for first mDNS address for app id %s", appID)
+		// wait for the context to be canceled or time out.
+		<-ctx.Done()
 
-	err := m.browse(ctx, appID, onFirst)
-	if err != nil {
-		return "", err
-	}
+		if errors.Is(ctx.Err(), context.Canceled) {
+			// expect this when we've found an address and canceled the browse.
+			m.logger.Debugf("Browsing for first mDNS address for app id %s canceled.", appID)
+		} else if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+			// expect this when we've been unable to find the first address before the timeout.
+			m.logger.Debugf("Browsing for first mDNS address for app id %s timed out.", appID)
+		}
 
-	// wait for the context to be canceled or time out.
-	<-ctx.Done()
+		// if onFirst has been invoked then we should have an address.
+		if addr == "" {
+			m.pubErrToSubs(appID, fmt.Errorf("couldn't find service: %s", appID))
 
-	if errors.Is(ctx.Err(), context.Canceled) {
-		// expect this when we've found an address and canceled the browse.
-		m.logger.Debugf("Browsing for first mDNS address for app id %s canceled.", appID)
-	} else if errors.Is(ctx.Err(), context.DeadlineExceeded) {
-		// expect this when we've been unable to find the first address before the timeout.
-		m.logger.Debugf("Browsing for first mDNS address for app id %s timed out.", appID)
-	}
+			done <- true // signal that all subscribers have been notified.
+			return
+		}
+
+		m.pubAddrToSubs(appID, addr)
 
-	if addr == "" {
-		return "", fmt.Errorf("couldn't find service: %s", appID)
+		done <- true // signal that all subscribers have been notified.
+	}()
+}
+
+func (m *resolver) pubErrToSubs(reqID string, err error) {
+	m.subMu.RLock()
+	defer m.subMu.RUnlock()
+	pool, ok := m.subs[reqID]
+	if !ok {
+		// we would always expect atleast 1 subscriber for this reqID.
+		m.logger.Warnf("no subscribers found for app id %s", reqID)
+		return
 	}
+	for _, subscriber := range pool.Subscribers {
+		// ErrChan is a buffered channel so this is non blocking.
+		subscriber.ErrChan <- err
+		subscriber.Close()
+	}
+}
 
-	return addr, nil
+func (m *resolver) pubAddrToSubs(reqID string, addr string) {
+	m.subMu.RLock()
+	defer m.subMu.RUnlock()
+	pool, ok := m.subs[reqID]
+	if !ok {
+		// we would always expect atleast 1 subscriber for this reqID.
+		m.logger.Warnf("no subscribers found for app id %s", reqID)
+		return
+	}
+	for _, subscriber := range pool.Subscribers {
+		// AddrChan is a buffered channel so this is non blocking.
+		subscriber.AddrChan <- addr
+		subscriber.Close()
```

Comment:
can this be called as part of the defer function in line 341?
As always a subscriber either receives error (even on cox deadline exceeded), or the address. In which case subscription channels can be closed once the subscriber consumes the address/error.

Reply:
It's safer to do this here where we are writing so we know the channel is open - if we do it in the defer statement there are other code paths such as the second cache check that will return before the sender may have written to the channel. In other words, there is a chance the channel may never be read but we want to guarantee that we can at least write to it. In fact, moving this to the defer does result in a panic trying to write to a closed channel during the test.